fairscale/experimental/nn/distributed_pipeline/trace.py

```diff
@@ -0,0 +1,101 @@
+import inspect
+import operator
+from typing import Dict, Optional, cast
+
+from torch.distributed.nn import RemoteModule
+import torch.fx
+import torch.nn as nn
+
+from . import PipelineModulesGraph
+
+
+class RemoteModuleTracer(torch.fx.Tracer):
+    def is_leaf_module(self, m: torch.nn.Module, module_qualified_name: str) -> bool:
+        if isinstance(m, RemoteModule):
+            return True
+        return False
+
+
+class GraphCreator:
+    def __init__(self, tracer: RemoteModuleTracer) -> None:
+        self.tracer = tracer
+
+    def get_module(self, node: torch.fx.Node) -> Optional[nn.Module]:
+        """Given a call_module node, returns the module corresponding to this module call"""
+        if node.op != "call_module":
+            return None
+        module = self.tracer.root
+        for t in cast(str, node.target).split("."):
+            module = getattr(module, t)
+        return module
+
+    def create_graph(self) -> PipelineModulesGraph:
+        node_to_data: Dict[torch.fx.Node, PipelineModulesGraph.DataSourceSpec] = {}
+        node_to_module = {}
+
+        for node in self.tracer.graph.nodes:
+            if node.op == "call_module":
+                module = self.get_module(node)
+                assert isinstance(module, RemoteModule)
+                node_to_data[node] = module
```

Comment:
i am not sure i understand why we assign both these structures to module.

Reply:
chose a better name to make it a bit more clear.